### ISSUE TITLE
Question: Dynamic registry and loading from proto sources

### ISSUE THREAD

**Issue by geoah** [jhump/protoreflect#163]
First of all I do apologise if I'm missing something very obvious or if I am misunderstanding or abusing things in this wonderful package.

I'm trying to setup a dynamic registry with a fetcher that is able to load protos from their source files without much success.

I've managed to get the dynamic registry to work when the fetcher manually constructs and returns a `*ptype.Type` but I'm struggling to figure out how to convert the result from the `protoparse.Parser` into a `*ptype.Type`. It seems to be dying trying to get the options, but I'm probably doing something very wrong here.

ps. Just a weird side-note, I cannot for the life of me debug this test using delve, it just dies with `Failed to produce stack trace!` -- will need to dig into this at some point :D

Thank you very much in advance for your help and also for this amazing package.
George

---

### Testcase

```golang
package dynamic

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"testing"

	"github.com/davecgh/go-spew/spew"
	"github.com/golang/protobuf/proto"
	"github.com/jhump/protoreflect/desc/protoparse"
	"github.com/jhump/protoreflect/dynamic"
	"github.com/jhump/protoreflect/dynamic/msgregistry"
	"github.com/stretchr/testify/assert"
)

func TestSomethingWeird2(t *testing.T) {
	protoSource := `
		syntax = "proto2";
		package foo;
		message Bar {
			optional string abc = 1;
			optional int32 def = 2;
		}`

	registry := (&msgregistry.MessageRegistry{}).
		WithMessageFactory(dynamic.NewMessageFactoryWithDefaults())

	reader := func(filename string) (io.ReadCloser, error) {
		if filename != "https://test/foo.Bar" {
			return nil, errors.New("not found")
		}
		return ioutil.NopCloser(bytes.NewReader([]byte(protoSource))), nil
	}

	parser := &protoparse.Parser{
		Accessor: reader,
	}

	fetcher := func(url string, enum bool) (proto.Message, error) {
		fmt.Println(url)
		if url != "https://test/foo.Bar" {
			return nil, errors.New("not found")
		}

		fds, err := parser.ParseFiles(url)
		if err != nil {
			return nil, err
		}

		if len(fds) == 0 {
			return nil, errors.New("missing descs")
		}

		fooMessage := fds[0].FindMessage("foo.Bar")
		spew.Dump(fooMessage)

		fooPType := registry.MessageAsPType(fooMessage)
		spew.Dump(fooPType)

		return fooPType, nil
	}

	registry = registry.WithFetcher(fetcher)

	md, err := registry.FindMessageTypeByUrl("test/foo.Bar")
	assert.NoError(t, err)
	assert.NotNil(t, md)

	if md == nil {
		t.FailNow()
	}

	dm := dynamic.NewMessageFactoryWithDefaults().NewMessage(md).(*dynamic.Message)
	dm.SetFieldByNumber(1, "fubar")
	dm.SetFieldByNumber(2, int32(123))
	dm.SetFieldByNumber(1, "snafu")
	dm.SetFieldByNumber(2, int32(456))

	js, err := dm.MarshalJSON()
	assert.NoError(t, err)
	fmt.Println(string(js))
}
```

### Result

```
=== RUN   TestSomethingWeird
(*desc.MessageDescriptor)(0xc0001ebb20)(name:"Bar" field:<name:"abc" number:1 label:LABEL_OPTIONAL type:TYPE_STRING json_name:"abc" > field:<name:"def" number:2 label:LABEL_OPTIONAL type:TYPE_INT32 json_name:"def" > )
--- FAIL: TestSomethingWeird2 (0.00s)
panic: reflect: call of reflect.Value.Type on zero Value [recovered]
	panic: reflect: call of reflect.Value.Type on zero Value

goroutine 19 [running]:
testing.tRunner.func1(0xc00014cb00)
	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:792 +0x387
panic(0x1436720, 0xc0002184c0)
	/usr/local/Cellar/go/1.11/libexec/src/runtime/panic.go:513 +0x1b9
reflect.Value.Type(0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.11/libexec/src/reflect/value.go:1713 +0x16c
nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry.(*MessageRegistry).options(0xc000149300, 0x1533bc0, 0x0, 0x10, 0x10, 0x149bea0)
	/Users/geoah/src/nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry/message_registry.go:596 +0x10c
nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry.(*MessageRegistry).fieldAsPType(0xc000149300, 0xc00020da40, 0x2)
	/Users/geoah/src/nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry/message_registry.go:456 +0x69
nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry.(*MessageRegistry).MessageAsPType(0xc000149300, 0xc0001ebb20, 0x1)
	/Users/geoah/src/nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry/message_registry.go:438 +0xcb
command-line-arguments.TestSomethingWeird2.func2(0xc0001e91c0, 0x14, 0x0, 0xc0001cef60, 0x30, 0x30, 0x1496840)
	/Users/geoah/src/nope.io/go/playground/dynamic/dynamic3_test.go:62 +0x33c
nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry.(*resolutionContext).addType(0xc000153dc0, 0xc0001e91c0, 0x14, 0xc0000bf9e0, 0x187b700, 0x0, 0x0)
	/Users/geoah/src/nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry/ptype_resolver.go:445 +0xbb
nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry.(*typeResolver).resolveUrlToMessageDescriptor(0xc000149308, 0xc0001e91c0, 0x14, 0x0, 0x0, 0x0)
	/Users/geoah/src/nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry/ptype_resolver.go:282 +0x278
nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry.(*MessageRegistry).FindMessageTypeByUrl(0xc000149300, 0x14b7940, 0xc, 0x105e370, 0xc00004c720, 0xe4b2a72b)
	/Users/geoah/src/nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry/message_registry.go:165 +0x91
command-line-arguments.TestSomethingWeird2(0xc00014cb00)
	/Users/geoah/src/nope.io/go/playground/dynamic/dynamic3_test.go:70 +0x21d
testing.tRunner(0xc00014cb00, 0x14cebd0)
	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:878 +0x353
FAIL	command-line-arguments	0.018s
```

**Comment by jhump** [comment]
Doh! You've found bugs in part of this repo where testing is very light.

So... I've added a test (based a bit on your example) and had to make numerous little fixes to get the test to pass. This will be fixed shortly (after #164 lands).

**Comment by geoah** [comment]
@jhump woa that was insanely fast -- thank you! :D
